Introduce remotecfg service (#6277)
grafana/agent: docs/sources/flow/reference/config-blocks/remotecfg.md

---
aliases:
- /docs/grafana-cloud/agent/flow/reference/config-blocks/remotecfg/
- /docs/grafana-cloud/monitor-infrastructure/agent/flow/reference/config-blocks/remotecfg/
- /docs/grafana-cloud/monitor-infrastructure/integrations/agent/flow/reference/config-blocks/remotecfg/
- /docs/grafana-cloud/send-data/agent/flow/reference/config-blocks/remotecfg/
canonical: remotecfgs://grafana.com/docs/agent/latest/flow/reference/config-blocks/remotecfg/
description: Learn about the remotecfg configuration block
menuTitle: remotecfg
title: remotecfg block
---

# remotecfg block

> `remotecfg` is an optional configuration block that enables {{< param "PRODUCT_NAME" >}}

Reviewer: Does this inherently mean that they cannot define other config blocks in the remote cfg? ie http,logging ect?

Reviewer 2: It cannot contain tracing and logging blocks, but it technically can contain argument and export blocks. I think that we should forbid it as it makes no sense to have them (the controller is marked as isolated)

Author: Hmm, that's a good point, should likely be changed.

> to fetch and load configuration from a remote endpoint.

Reviewer 3: ```suggestion
to fetch and load the configuration from a remote endpoint.
```

`remotecfg` is specified without a label and can only be provided once per
configuration file.

The [API definition][] for managing and fetching configuration that the
`remotecfg` block uses is available under the Apache 2.0 license.

[API definition]: https://github.com/grafana/agent-remote-config

## Example

```river
remotecfg {
	url = "SERVICE_URL"
	basic_auth {
		username      = USERNAME
		password_file = PASSWORD_FILE
	}

	id             = constants.hostname
	metadata       = {"cluster" = "dev-us-central-0", "namespace" = "agent-otlp"}
	poll_frequency = "10m"
}
```

## Arguments

The following arguments are supported:

Name             | Type                 | Description                                      | Default     | Required
-----------------|----------------------|--------------------------------------------------|-------------|---------

> `url`            | `string`             | The address of the API to poll for configuration | `""`        | no

Reviewer 4: suggestion: this is marked as not required because the service itself isn't required for the agent to run, right? But for the service to function, a URL does need to be provided. Perhaps there should be a note below the table that the URL _is_ functionally required if the service is to run.

Author: Yes; right now services blocks are always implicitly constructed (so there's no way to have an required field).

We can expand on this with a note below.

> `url`            | `string`             | The address of the API to poll for configuration | `""`        | no

Reviewer 3: ```suggestion
`url`            | `string`             | The address of the API to poll for configuration. | `""`        | no
```

`id`             | `string`             | A self-reported ID                               | `see below` | no
`metadata`       | `map(string)`        | A set of self-reported metadata                  | `{}`        | no

> `poll_frequency` | `duration`           | How often to poll the API for new configuration  | `"1m"       | no

Reviewer 4: ```suggestion
`poll_frequency` | `duration`           | How often to poll the API for new configuration  | `"1m"`      | no
```

> `id`             | `string`             | A self-reported ID                               | `see below` | no
> `metadata`       | `map(string)`        | A set of self-reported metadata                  | `{}`        | no
> `poll_frequency` | `duration`           | How often to poll the API for new configuration  | `"1m"`      | no

Reviewer 3: ```suggestion
`id`             | `string`             | A self-reported ID.                               | `see below` | no
`metadata`       | `map(string)`        | A set of self-reported metadata.                  | `{}`        | no
`poll_frequency` | `duration`           | How often to poll the API for new configuration.  | `"1m"`      | no
```

> If not set, the self-reported `id` that the Agent uses is a randomly generated,

Reviewer 3: ```suggestion
If not set, the self-reported `id` that {{< param "PRODUCT_ROOT_NAME" >}} uses is a randomly generated,
```

> If not set, the self-reported `id` that the Agent uses is a randomly generated,
> anonymous unique ID (UUID) that is stored on the Agent's storage path as
> `agent_seed.json` so that it can persist across restarts.

Reviewer 4: nit:
```suggestion
If not set, the self-reported `id` that the Agent uses is a randomly generated,
anonymous unique ID (UUID) that is stored in the Agent's storage path as
`agent_seed.json` so that it can persist across restarts.
```

Reviewer 4: Actually looking at this a second time this may not be any clearer? :thinking:

Reviewer 3: We should not call it `the Agent`. It's a collector. Call it by its name: {{< param "PRODUCT_ROOT_NAME" >}} for Grafana Agent or {{< param "PRODUCT_NAME" >}} for Grafana Agent Flow.

> anonymous unique ID (UUID) that is stored on the Agent's storage path as
> `agent_seed.json` so that it can persist across restarts.

Reviewer 3: ```suggestion
anonymous unique ID (UUID) that is stored on t{{< param "PRODUCT_ROOT_NAME" >}}'s storage path as
`agent_seed.json` so that it can persist across restarts.
```
